Image Optimization: Bulk Image Optimization

arunshenoy99 opened the pull request:
## Proposed changes

<!-- Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. -->

## Type of Change

<!-- What types of changes does your code introduce? -->
<!-- _Put an `x` in the boxes that apply_ -->

#### Production

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update
- [ ] Refactoring / housekeeping (changes to files not directly related to functionality)

<!-- Bugfixes should explain how to reproduce the bug -->

#### Development

- [ ] Tests
- [ ] Dependency update
- [ ] Environment update / refactoring
- [ ] Documentation Update

<!-- All PRs should endeavor to include tests -->
<!-- PRs with new tools or dependencies should explain how to use them -->

## Video

<!-- Add a short video demonstrating where the change takes effect and how to can be reproduced by reviewers -->
<!-- On macOS press cmd+shift+5 to open the screen recording tool. It will save the video to the desktop  -->

<!-- Drag and drop the video file into the GitHub editor to attach it -->

## Checklist

<!-- _Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._ -->

- [ ] I have read the [CONTRIBUTING](https://github.com/bluehost/.github/blob/master/.github/contributing.md) doc
- [ ] I have viewed my change in a web-browser
- [ ] Linting and tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

<!-- 
## Pre-deployment Checklist

- [ ] E.g. environmental variables which need to be set before deployment will work.
- [ ] E.g. other tickets which need to be complete before an API this change relies on will be ready
-->

<!--
## Post-deployment Checklist

How can the change be verified?

- [ ] E.g. temporarily enable logging and observe specific logs.
- [ ] E.g. observe new data in a specific view or table.
-->

## Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->
<!-- Now is a good time to create additional tickets for any necessary follow-up work. -->

bradp commented on scripts/bulk-optimizer.js:
```diff
@@ -0,0 +1,269 @@
+
+		const modalTitle = document.createElement( 'h2' );
+		modalTitle.id = 'nfd-modal-title';
+		modalTitle.textContent = 'Optimizing Images...';
```
🔢 Localize text please.

bradp commented on scripts/bulk-optimizer.js:
```diff
@@ -0,0 +1,269 @@
+
+		const currentFileName = document.createElement( 'p' );
+		currentFileName.id = 'nfd-current-file';
+		currentFileName.textContent = 'Preparing files...';
```
🔢 Localize text please.

bradp commented on includes/Images/ImageSettings.php:
```diff
@@ -83,6 +84,11 @@ private function register_settings() {
 									),
 								),
 							),
+							'bulk_optimization' => array(
```
Can we move this option above lazy loading to have it be next to the other optimization ones?

geckod22 commented on scripts/bulk-optimizer.js:
```diff
@@ -0,0 +1,269 @@
+		try {
+			for ( let i = 0; i < selectedItems.length; i++ ) {
+				if ( cancelRequested ) {
+					modalTitle.textContent = 'Optimization Canceled';
```
Localize text

geckod22 commented on scripts/bulk-optimizer.js:
```diff
@@ -0,0 +1,269 @@
+				}
+
+				const { id: mediaId, name: fileName } = selectedItems[ i ];
+				currentFileName.textContent = `Optimizing: ${ fileName }`;
```
Localize string

geckod22 commented on scripts/bulk-optimizer.js:
```diff
@@ -0,0 +1,269 @@
+			}
+
+			if ( ! cancelRequested ) {
+				modalTitle.textContent = 'Optimization Complete!';
```
Localize string

geckod22 commented on scripts/bulk-optimizer.js:
```diff
@@ -0,0 +1,269 @@
+		bulkOptimizeButton.id = bulkOptimizeButtonId;
+		bulkOptimizeButton.className =
+			'button media-button button-large button-primary';
+		bulkOptimizeButton.textContent = 'Optimize';
```
Localize string

geckod22 commented on scripts/bulk-optimizer.js:
```diff
@@ -0,0 +1,269 @@
+				setTimeout( closeModal, 2000 );
+			}
+		} catch ( error ) {
+			modalTitle.textContent = 'An error occurred.';
```
Localize string

geckod22 commented on scripts/bulk-optimizer.js:
```diff
@@ -0,0 +1,269 @@
+			document.querySelectorAll( '.attachment.selected' )
+		).map( ( attachment ) => ( {
+			id: attachment.getAttribute( 'data-id' ),
+			name: getFileName( attachment ),
```
it seems filename is always the same if we have more than one image selected because in getFileName function on 134 line where the file name is get by the .media-frame-content .filename that only provide the current selected file, so in case of more selected we only get the last seletec element file name.

 Maybe we could change the line  like this

```suggestion
			name: attachment.getAttribute( 'aria-label' ),
```
and then remove the getFileName function on line 133

arunshenoy99 replied:
Thanks! @geckod22 updated.

arunshenoy99 commented:
> About the bulk operation in media gallery I was also thinking this: with multiple selected files at the end of the process if some of the images gave an error, the user really will never konw as the modal of course will update with the information of the following file and at the end will set as Operation complete and then will close. Should we maybe add a little banner like the result of the process (only in case with multiple selection). IT is just an idea, what do you think? maybe we don't want to give this feedback.

I added a simple results screen to the modal. I plan to create a final PR with all the merged changes and will tweak the design in that one:
<img width="487" alt="image" src="https://github.com/user-attachments/assets/6fdd01ca-85e1-4ed0-be0b-1032e667084d" />

arunshenoy99 commented:
I have addressed all the comments. I am just merging this PR into its base branch. I’ll put together a combined PR for all the Image Optimization changes so we can see the full picture and make reviewing easier.
